# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x + 2

print(addTwo(4))

# Lambda function syntax
a_var = 4
print((lambda x: x+2)(a_var))

def addTwo(x):
    return x + 2

print(addTwo(4))

print(lambda x: x+2(4))



6
6
6
<function <lambda> at 0x00000211E1EA1BC0>


<>:15: SyntaxWarning: 'int' object is not callable; perhaps you missed a comma?
<>:15: SyntaxWarning: 'int' object is not callable; perhaps you missed a comma?
C:\Users\Godzilla\AppData\Local\Temp\ipykernel_1924\3726679416.py:15: SyntaxWarning: 'int' object is not callable; perhaps you missed a comma?
  print(lambda x: x+2(4))


#### Saving to a Variable

In [2]:


test = lambda x: x+2
test(4)

6

#### Multiple Inputs

In [3]:
# Multiple inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

x_test = lambda x,y,z: x*y*z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [4]:
def multiply(f, num):
    """
    f expects a lambda function
    num expects a number
    """
    return f(num)
multiply(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [9]:
def multiply_test(num):
    return num*4

# Function in a function

def returnFunc():
    test = 4
    def multiply(num):
        return num*2
    return multiply

f_return = returnFunc()
print(f_return)
print(f_return(4))

# Lambda function return from a regular function
def returnLam(b,c):
    return lambda x,a: x+a+b+c
r_lamb = returnLam(4,5)
print(r_lamb)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x00000211E1EA27A0>
8
<function returnLam.<locals>.<lambda> at 0x00000211E1EA2DE0>
19


#### If Statements within Lambdas

In [10]:
# Lambda x: True if (condition) else False
f_condition = lambda num:num *2 if num >10 else num+2

print(f_condition(8))
print(f_condition(20))

10
40


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [12]:
print(lambda x: x**4)


<function <lambda> at 0x00000211E1F34720>


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [8]:
"""
map(function,iterable)
an iterable is something you can loop over; such as a list, dict, tuple, set orstring
Normally, the usage of map happens with a pre-defined function, but we can use lambdas as well
"""
def squared(num, num2):
    if num <10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2
numbers = [4,11,20,3,15,20]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [14]:
# Define the iterable
numbers = [1, 2, 3, 4, 5]
more_nums = [6, 7, 8, 9, 10]

# Use map function with lambda expression
squared_nums_lamb = list(map(lambda x: (x[0]**2, x[1]**2) if x[0] < 10 and x[1] < 10 else x, zip(numbers, more_nums)))
print(squared_nums_lamb)


[(1, 36), (4, 49), (9, 64), (16, 81), (5, 10)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [15]:
formula1 = list(map(lambda x: 2*x-1, numbers))
print(formula1)

formula2 = list(map(lambda x: x**3-20, numbers))
print(formula2)

[1, 3, 5, 7, 9]
[-19, -12, 7, 44, 105]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [17]:
# list(filter(function, iterable))
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else: 
        return False

new_names_list = list(filter(a_names,names))
print(new_names_list)

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [18]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [22]:
import statistics
numbers = [1, 2, 3, 4, 5]
#find the mean
mean = statistics.mean(numbers)
#lambda of x
below = list(filter(lambda x: x < mean, numbers))
print(below)

[1, 2]


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [23]:
def add_nums(num):
    # Set a base case to stop recursion from going for forever!!!
    if num <= 1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({num - 1})')
    return num + add_nums(num-1)

add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [24]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num-1)

factorial(5)

120

#### In-Class Exercise #4 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [27]:

def sub_nums(num):
    #base case
    if num <= 1:
        print('sub_nums(1)= 1')
        return num
    else:
        print(f'sub_nums{num} = {num} - sub_nums(num - 1)')
    return num - sub_nums(num-1)

print(sub_nums)

<function sub_nums at 0x00000211E2477880>


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [25]:
def my_range(stop,start=0, step=2):
    while start < stop:
        yield start
        start += step
        
for i in my_range(10):
    print(i)
print([i for i in my_range(10)])

0
2
4
6
8
[0, 2, 4, 6, 8]


#### In-Class Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [29]:



def my_rangetwo(end, begin=0, move=2):
    while begin < end:
        yield begin **2
        begin += move
for i in my_rangetwo(10):
    print(i)
print([i for i in my_rangetwo(10)])
        

0
4
16
36
64
[0, 4, 16, 36, 64]


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [34]:

places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]



def filter_places(city):
    if city == "" or city.isspace():
        return False
    else:
        return True

new_list = list(filter(filter_places, places))
print(new_list)




['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [30]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`


In [35]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
places = [('Nashua', 32), ("Boston", 12), ("Los Angelos", 44), ("Miami", 29)]

converted_places = list(map(lambda x: (x[0], (9/5)*x[1]+32), places))

print(converted_places)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [36]:
def fib(n):
    if n == 0 or n == 1:
        return 1
    else:
        return fib(n-1) + fib(n-2)

for i in range(6):
    print(f"Iteration {i}: {fib(i)}")


Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
